packages/aws-cdk-lib/aws-cloudfront/lib/realtime-log-config.ts

```diff
@@ -0,0 +1,144 @@
+import { Construct } from 'constructs';
+import { CfnRealtimeLogConfig } from './cloudfront.generated';
+import * as iam from '../../aws-iam';
+import * as kinesis from '../../aws-kinesis';
+import { IResource, Names, PhysicalName, Resource } from '../../core';
+
+/**
+ * @internal
+ */
+function singletonKinesisRole(scope: Construct): iam.IRole {
+  const id = 'RealtimeLogKinesisRole';
+  const existing = scope.node.tryFindChild(id) as iam.IRole;
+  if (existing) { return existing; }
+
+  const role = new iam.Role(scope, id, {
+    roleName: PhysicalName.GENERATE_IF_NEEDED,
+    assumedBy: new iam.ServicePrincipal('cloudfront.amazonaws.com'),
+  });
+
+  return role;
+}
+
+/**
+ * Represents the endpoints available for targetting within a realtime log config resource
+ */
+export abstract class Endpoint {
+  /**
+   * Configure a Kinesis Stream Endpoint for Realtime Log Config
+   *
+   * @default - a role will be created and used across your endpoints
+   */
+  public static fromKinesisStream(scope: Construct, stream: kinesis.IStream, role?: iam.IRole): Endpoint {
+    const cloudfrontRole = role || singletonKinesisRole(scope);
+
+    if (!role) {
+      stream.grant(cloudfrontRole,
+        'kinesis:DescribeStreamSummary',
+        'kinesis:DescribeStream',
+        'kinesis:PutRecord',
+        'kinesis:PutRecords',
+      );
+
+      if (stream.encryptionKey) {
+        stream.encryptionKey.grant(cloudfrontRole, 'kms:GenerateDataKey');
+      }
+    }
+
+    return new (class extends Endpoint {
+      public _renderEndpoint() {
+        return {
+          kinesisStreamConfig: {
+            roleArn: cloudfrontRole.roleArn,
+            streamArn: stream.streamArn,
+          },
+          streamType: 'Kinesis',
+        };
+      }
+    });
+  }
+
+  private constructor() {}
+
+  /**
+  * @internal
+  */
+  public abstract _renderEndpoint(): any;
```

Comment:
How about this:

```ts
  public abstract _renderEndpoint(scope: Construct): any;
```

which let's you do `endpoint._renderEndpoint(this)` when creating the `CfnRealtimeLogConfig` (the role is created in the scope of the realtime log config) and then you don't have to pass a scope when using `fromKinesisStream()`

Reply:
Makes sense - will change!